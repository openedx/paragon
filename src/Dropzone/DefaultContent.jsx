import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage, useIntl } from 'react-intl';
import { getTypesString, formatBytes } from './utils';
import messages from './messages';
import Icon from '../Icon';
import { FileUpload } from '../../icons';

const DefaultContent = ({ accept, minSize, maxSize }) => {
  const intl = useIntl();

  const getFileRestrictionMessage = () => {
    let fileTypePart;
    let fileSizePart;

    if (accept) {
      const allTypes = getTypesString(accept);
      const lastTypeSeparatorLocation = allTypes.lastIndexOf(',');
      fileTypePart = intl.formatMessage(messages.fileTypeRestriction, {
        count: lastTypeSeparatorLocation === -1 ? 1 : 2,
        firstPart: lastTypeSeparatorLocation === -1 ? allTypes : allTypes.slice(0, lastTypeSeparatorLocation),
        secondPart: lastTypeSeparatorLocation !== -1 && allTypes.slice(lastTypeSeparatorLocation + 1),
      });
    }

    if (minSize && maxSize && Number.isFinite(maxSize)) {
      fileSizePart = intl.formatMessage(messages.fileSizeBetween, {
        sizeMin: formatBytes(minSize),
        sizeMax: formatBytes(maxSize),
      });
    } else if (maxSize && Number.isFinite(maxSize)) {
      fileSizePart = intl.formatMessage(messages.fileSizeMax, { sizeMax: formatBytes(maxSize) });
    } else if (minSize) {
      fileSizePart = intl.formatMessage(messages.fileSizeMin, { sizeMin: formatBytes(minSize) });
    }

    if (fileTypePart && fileSizePart) {
      return `${fileTypePart} (${fileSizePart})`;
    }
    if (fileTypePart) {
      return fileTypePart;
    }
    return fileSizePart;
  };

  return (
    <>
      <div className="pgn__dropzone__upload-icon-wrapper">
        <Icon src={FileUpload} className="pgn__dropzone__upload-icon" />
      </div>
      <p>
        <FormattedMessage
          id="pgn.Dropzone.DefaultContent.label"
          description="A text that appears as a label for input of Dropzone component."
          defaultMessage="Drag and drop your file here or click to upload."
        />
      </p>
      {[accept, minSize, maxSize].some(value => value) && (
        <p className="x-small text-gray-500">
          {getFileRestrictionMessage()}
        </p>
      )}
    </>
  );
};

DefaultContent.propTypes = {
  accept: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)),
  maxSize: PropTypes.number,
  minSize: PropTypes.number,
};

DefaultContent.defaultProps = {
  accept: undefined,
  maxSize: undefined,
  minSize: undefined,
};

export default DefaultContent;
