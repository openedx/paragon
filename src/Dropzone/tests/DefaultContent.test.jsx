import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

import DefaultContent from '../DefaultContent';

// eslint-disable-next-line react/prop-types
function DefaultContentWrapper({ children, ...props }) {
  return (
    <IntlProvider locale="en" messages={{}}>
      <DefaultContent {...props} />
    </IntlProvider>
  );
}

describe('<Dropzone.DefaultContent />', () => {
  it('successfully renders', () => {
    const tree = renderer.create(<DefaultContentWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders file type restriction message if receives accept prop', () => {
    const { container } = render(<DefaultContentWrapper accept={{ 'image/*': ['.png', '.jpg'] }} />);
    expect(container).toHaveTextContent('Upload PNG or JPG files');
  });
  it('renders min size restriction message if receives minSize prop', () => {
    const { container } = render(<DefaultContentWrapper minSize={1048576} />);
    expect(container).toHaveTextContent('Min 1MB');
  });
  it('renders max size restriction message if receives maxSize prop', () => {
    const { container } = render(<DefaultContentWrapper maxSize={1048576} />);
    expect(container).toHaveTextContent('Max 1MB');
  });
  it('renders type and size restriction message if receives all props', () => {
    const { container } = render(
      <DefaultContentWrapper
        accept={{ 'image/*': ['.png', '.jpg', '.gif'] }}
        minSize={1048576}
        maxSize={20 * 1048576}
      />,
    );
    expect(container).toHaveTextContent('Upload PNG, JPG or GIF files (Between 1MB and 20MB)');
  });
});
