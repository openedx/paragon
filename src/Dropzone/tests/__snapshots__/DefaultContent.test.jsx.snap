// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropzone.DefaultContent /> successfully renders 1`] = `
Array [
  <div
    className="pgn__dropzone__upload-icon-wrapper"
  >
    <span
      className="pgn__icon pgn__dropzone__upload-icon"
    >
      <svg
        aria-hidden={true}
        fill="none"
        focusable={false}
        height={24}
        role="img"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 10h4v6h6v-6h4l-7-7-7 7zm0 8v2h14v-2H5z"
          fill="currentColor"
        />
      </svg>
    </span>
  </div>,
  <p>
    Drag and drop your file here or click to upload.
  </p>,
]
`;
